### ISSUE TITLE
Logging to aws AES cluster may fails with 413 errors (Request Entity Too Large)

### ISSUE THREAD

**Issue by fredleger** [rancher/rancher#21610]
**What kind of request is this (question/bug/enhancement/feature request):** bug

**Steps to reproduce (least amount of steps as possible):** 
* deploy rancher 2.2.4 on aws
* deploy an aws elasticsearch cluster with t2.small.elasticsearch, one AZ and only one instance
* create an external service pointing to the ES cluster
* configure rancher to use this service
*  wait a bit to have some logs

**Result:**

We can see errors in fluentd like this one: 

```
2019-07-18 09:04:23 +0000 [warn]: #0 failed to flush the buffer. retry_time=10 next_retry_seconds=2019-07-18 09:13:55 +0000 chunk="58dada07f4f4df922debb60794495f09" error_class=Elasticsearch::Transport::Transport::Errors::RequestEntityTooLarge error="[413] {\"Message\":\"Request size exceeded 10485760 bytes\"}"
```

**Other details that may be helpful:**
https://docs.aws.amazon.com/elasticsearch-service/latest/developerguide/aes-limits.html

**Environment information**
- Rancher version (`rancher/rancher`/`rancher/server` image tag or shown bottom left in the UI): 2.2.4
- Installation option (single install/HA): single (inside the k8s cluster via helm)

**Cluster information**
- Cluster type (Hosted/Infrastructure Provider/Custom/Imported): Imported (created via kops)
- Machine type (cloud/VM/metal) and specifications (CPU/memory): cloud / aws / t3.medium
- Kubernetes version (use `kubectl version`): 1.12.8

Is there any way we can control the size of the payload here ? if yes i think this should be presented ass a configuration option when deploying rancher logging stack

**Comment by gillarda** [comment]
I am having the same issue in 2.2.6.
Unless I am mistaken I did not have this behavior in 2.2.3. 
It seems there has been a lot of work between 2.2.3 and 2.2.4 (notably #20208 and #20047). These improved usability a lot but may have introduced this behavior.
Digging a bit I also found a commit to [rancher/fluentd](https://github.com/rancher/fluentd/commit/f137dccf2ffd65e096ea387305d6aaaa92a6f769) referencing #14987 that updated the fluentd version to address an issue with a lot of buffer files being generated.
Maybe @aiwantaozi has an idea ?

**Comment by cjellick** [comment]
@loganhz @aiwantaozi please investigate and determine if this is a regression

**Comment by aiwantaozi** [comment]
Hi @fredleger Just check the AES doc, the max size of http request payloads of t2.small is 10M. Seems like you can't edit it in AWS, more details could reach here.  https://docs.aws.amazon.com/elasticsearch-service/latest/developerguide/aes-limits.html
![image](https://user-images.githubusercontent.com/5936271/63261426-0b3e9d00-c2b6-11e9-90d9-7485ecc0de5b.png)

You could use advanced mode to input the configure, then you could add a threshold configure `bulk_message_request_threshold 10485760` to match the AES. `10485760 = 10 * 1024 * 1024`
Step:
1. `Tools` -> `Logging` -> click target `Elasticsearch` -> `Edit as File`
2. input your configure, sample as below, more configure could reach here https://github.com/uken/fluent-plugin-elasticsearch
![image](https://user-images.githubusercontent.com/5936271/63261740-c6ffcc80-c2b6-11e9-99fc-c4ceb75a3713.png)

**Comment by gillarda** [comment]
Thank you,

I tried what you suggested but got the logs below returned by the  `rancher-logging-fluentd` pods

Rancher 2.2.6
rancher-logging 0.1.1
rancher/fluentd:v0.1.13

```
2019-08-19 12:19:17 +0000 [warn]: parameter 'bulk_message_request_threshold' in <store>
8/19/2019 2:19:17 PM   @type "elasticsearch"
8/19/2019 2:19:17 PM   include_tag_key true
8/19/2019 2:19:17 PM   hosts "https://aes.es.amazonaws.com:443"
8/19/2019 2:19:17 PM   logstash_prefix "es-index"
8/19/2019 2:19:17 PM   logstash_format true
8/19/2019 2:19:17 PM   logstash_dateformat "%Y-%m-%d"
8/19/2019 2:19:17 PM   type_name "container_log"
8/19/2019 2:19:17 PM   ssl_verify false
8/19/2019 2:19:17 PM   ssl_version TLSv1_2
8/19/2019 2:19:17 PM   bulk_message_request_threshold 10485760
8/19/2019 2:19:17 PM   slow_flush_log_threshold 40.0
8/19/2019 2:19:17 PM   <buffer>
8/19/2019 2:19:17 PM     @type "file"
8/19/2019 2:19:17 PM     path "/fluentd/log/buffer/project.c-iudrf_p-k7gsq.buffer"
8/19/2019 2:19:17 PM     flush_mode interval
8/19/2019 2:19:17 PM     flush_interval 60s
8/19/2019 2:19:17 PM     flush_thread_count 8
8/19/2019 2:19:17 PM     queued_chunks_limit_size 200
8/19/2019 2:19:17 PM   </buffer>
8/19/2019 2:19:17 PM </store> is not used.
```

**Comment by loganhz** [comment]
@aiwantaozi please look at it. But it's not a regression.

**Comment by aiwantaozi** [comment]
@gillarda bulk_message_request_threshold this configure is available after fluent-plugin-elasticsearch v3.5.1, we upgrade to v3.5.4, but system-chart does not release yet, you can try upgrade the images rancher/fluentd:v0.1.14
Steps: 
1. System project -> Apps -> Rancher-Logging -> Upgrade it
2. Input answer `fluentd.image.tag` = `v0.1.14`
3. finished upgrade then configure `bulk_message_request_threshold` should be available

**Comment by fredleger** [comment]
@aiwantaozi yes it was clear that was linked to the AWS limitations here but the fisrt setup that allow 100Mo is quite expensive actually.

The point is more that there is not easy setting to handle in the rancher chart and i think this could be a improvement.

Also 10MB (or a bit less see bellow) could  be a sane default value what do you think ?

I had to tweak it a bit to make it really work :

```
<match *>
  @type elasticsearch
  include_tag_key true
  hosts https://my-es-managed.eu-west-3.es.amazonaws.com:443
  logstash_prefix "demo"
  logstash_format true
  logstash_dateformat %Y-%m-%d
  type_name "container_log"
  ssl_verify false
  ssl_version TLSv1_2
  bulk_message_request_threshold 9437184
  request_timeout 240s
</match>
```

My feeling here is that 10MB as a settings is the size of the payload and there is a bit of extra data added afterwards so lowering it made it work.

@gillarda did you try that as well ?

**Comment by aiwantaozi** [comment]
@fredleger Thank you for letting us know, really helpful.

**Comment by gillarda** [comment]
It works for me as well even though my log pattern changed a bit so I cannot exactly reproduce the context. However, 15MB or 20MB buffer files get processed without issue.

I got it to work with a 10MB value, but I guess if there is some compression/uncompression involved (or some extra data added as suggested), the threshold has to be lower.

I agree that adding a few more options to the GUI would be a great improvement.

Anyway, thank you for the help !
